File: ordered_map.go
package iriq

// OrderedMap is an insertion-ordered string->string map. We mirror Ruby's Hash
// ordering semantics for query parameters and other fields where users see the
// declared order (and where JSON round-trips need to be stable).
type OrderedMap struct {
	keys   []string
	values map[string]string
}

func NewOrderedMap() *OrderedMap {
	return &OrderedMap{values: map[string]string{}}
}

func (m *OrderedMap) Len() int {
	if m == nil {
		return 0
	}
	return len(m.keys)
}

func (m *OrderedMap) Set(k, v string) {
	if _, ok := m.values[k]; !ok {
		m.keys = append(m.keys, k)
	}
	m.values[k] = v
}

func (m *OrderedMap) Get(k string) (string, bool) {
	if m == nil {
		return "", false
	}
	v, ok := m.values[k]
	return v, ok
}

func (m *OrderedMap) Keys() []string {
	if m == nil {
		return nil
	}
	out := make([]string, len(m.keys))
	copy(out, m.keys)
	return out
}

// Each iterates in insertion order.
func (m *OrderedMap) Each(fn func(k, v string)) {
	if m == nil {
		return
	}
	for _, k := range m.keys {
		fn(k, m.values[k])
	}
}

func (m *OrderedMap) ToMap() map[string]string {
	if m == nil {
		return map[string]string{}
	}
	out := make(map[string]string, len(m.values))
	for k, v := range m.values {
		out[k] = v
	}
	return out
}

File: rust/iriq/src/ordered_map.rs
use std::collections::HashMap;

/// Insertion-ordered string → string map. Mirrors Ruby Hash ordering for
/// query parameters and other fields where users see the declared order.
///
/// Values are `Option<String>` because Ruby distinguishes a valueless param
/// ("?flag" → nil) from an empty one ("?flag=" → ""). Downstream consumers
/// treat both as "" (Ruby calls `.to_s`); only the raw parse dump surfaces
/// the difference — use `iter_raw` there.
#[derive(Debug, Clone, Default)]
pub struct OrderedMap {
    keys: Vec<String>,
    values: HashMap<String, Option<String>>,
}

impl OrderedMap {
    pub fn new() -> Self {
        Self::default()
    }

    pub fn len(&self) -> usize {
        self.keys.len()
    }

    pub fn is_empty(&self) -> bool {
        self.keys.is_empty()
    }

    pub fn set(&mut self, k: impl Into<String>, v: Option<String>) {
        let key = k.into();
        if !self.values.contains_key(&key) {
            self.keys.push(key.clone());
        }
        self.values.insert(key, v);
    }

    pub fn get(&self, k: &str) -> Option<&str> {
        self.values.get(k).and_then(|v| v.as_deref())
    }

    pub fn keys(&self) -> Vec<String> {
        self.keys.clone()
    }

    /// Values with nil collapsed to "" — Ruby's `.to_s` view.
    pub fn iter(&self) -> impl Iterator<Item = (&str, &str)> {
        self.iter_raw().map(|(k, v)| (k, v.unwrap_or("")))
    }

    /// Values as parsed: None for "?flag", Some("") for "?flag=".
    pub fn iter_raw(&self) -> impl Iterator<Item = (&str, Option<&str>)> {
        self.keys
            .iter()
            .map(move |k| (k.as_str(), self.values.get(k).and_then(|v| v.as_deref())))
    }
}
